To int to nat removal semantics: RISCV Merge [1/3]

salinhkuhn opened the pull request:
This is a rather big PR again but it contains the semantics already merged and additionally has rewrites of the semantics into purely bit vector domain statements without calls to toInt and toNat. This file should replace the current smenatics file when complete.

This PR is the first third of the whole undertaking.
For some theorems, we previously already committed their pure variant but left out the toNat version. In this case, we now add the toNat version under the original theorem name, move the pure version to a _bv statement, and prove equivalence of both.

Commit:
first commit, added definintions

Commit:
added comment

Commit:
removed a comment

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,645 @@
+-/
+
+/-! ## Helper Bit Vector Lemmas  -/
+/- The lemmas bellow are helper lemmas used within proofs throughout the file -/
```
These all should probably go into `forLean.lean`.

tobiasgrosser replied:
```suggestion
/- The lemmas below are helper lemmas used within proofs throughout the file. -/
```

Please do a spell-check and terminate your sentences with `.`.

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,645 @@
+  rw [BitVec.signExtend_eq_setWidth_of_le _ (by omega)]
+  rw [BitVec.signExtend_eq_setWidth_of_le _ (by omega)]
+
+
```
```suggestion

```
No need for two empty lines in a row.

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,645 @@
+
+
+@[simp] -- we tag the lemma as a simp lemma
+theorem extractLsb'_eq_setWidth {x : BitVec w} : x.extractLsb' 0 n = x.setWidth n := by
```
```suggestion
@[simp]
theorem extractLsb'_eq_setWidth {x : BitVec w} : x.extractLsb' 0 n = x.setWidth n := by
```

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,645 @@
+    simp only [BitVec.ofInt_toInt, extractLsb'_eq_setWidth]
+    rw [extractLsb_setWidth_of_lt (hi := 127) (lo := 64) (v := 128) (x := BitVec.signExtend 129 rs1_val * BitVec.signExtend 129 rs2_val) (by omega)]
+    simp
+-- wait until accepted then add divison etc
```
```suggestion

```

tobiasgrosser commented:
Excellent. The overall bitvector level theorems look great, but I feel they could overall benefit from some polishing. Maybe @luisacicolini can help with a review here?

salinhkuhn committed:
Update SSA/Projects/RISCV64/Semantics2.lean

Co-authored-by: Tobias Grosser <tobias@grosser.es>

salinhkuhn committed:
Update SSA/Projects/RISCV64/Semantics2.lean

Co-authored-by: Tobias Grosser <tobias@grosser.es>

salinhkuhn committed:
Update SSA/Projects/RISCV64/Semantics2.lean

Co-authored-by: Tobias Grosser <tobias@grosser.es>

salinhkuhn committed:
Update SSA/Projects/RISCV64/Semantics2.lean

Co-authored-by: Tobias Grosser <tobias@grosser.es>

luisacicolini commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,642 @@
+not contain BitVec.toInt and BitVec.toNat. This file should replace
+the Semantics file in RISCV64 when complete.
+ToInt/ToNat calls hinder proof automation by bv_decide.
+To guarantee faithfullness to the semantics defined by the Sail-Lean RiscV64 model,
```
```suggestion
faithfulness
```
typo

luisacicolini commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,642 @@
+theorem sshiftRight_eq_sshiftRight_extractLsb {w : Nat}
+    {lw : Nat} (hll : 2^lw = w) (hlw : lw > 0 )
+    (x y : BitVec w)(h : BitVec.toNat y < 2^lw ) :
+        x.sshiftRight y.toNat = x.sshiftRight (y.extractLsb (lw - 1) 0).toNat := by
```
```suggestion
theorem sshiftRight_eq_sshiftRight_extractLsb {w : Nat}
    {lw : Nat} {x y : BitVec w} (hll : 2^lw = w) (hlw : lw > 0)
    (h : BitVec.toNat y < 2^lw) :
        x.sshiftRight y.toNat = x.sshiftRight (y.extractLsb (lw - 1) 0).toNat := by
```
small formatting suggestion: we try to follow lean's guidelines and style, and in particular: 
* avoid spurious spaces
* call the "arguments" implicitly when possible (using the curly braces, unless this breaks the statement) and having first the implicit ones and then the hypotheses

could you check that this is the case for the remaining theorems as well? :)

luisacicolini commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,642 @@
+    (x y : BitVec w)(h : BitVec.toNat y < 2^lw ) :
+        x.sshiftRight y.toNat = x.sshiftRight (y.extractLsb (lw - 1) 0).toNat := by
+  apply BitVec.eq_of_toNat_eq
+  simp
```
```suggestion
  simp
```
For this we follow lean's strategy too: having `simp` in the middle of proofs could be problematic, in fact the theorem that the tactic considers are constantly changing, and if we use it like this it might be that in subsequent updates to lean we rely on this code will break because of changes to the `simp` tactic. Therefore, we need to be sure that we know _which_ theorems `simp` actually uses and replace `simp` with `simp only`. 
The only exception to this is when the tactic is at the _end_ of the theorem (terminal simp) or followed by `omega` only, in which case there is no need to use `simp only`.
To find the theorems that are used, you can type `simp?` and apply that suggestion. 
Could you replace all the non-terminal `simp` with `simp only`?

luisacicolini commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,642 @@
+    rw [← this ]
+  have : (lw - 1 + 1) = lw := by omega
+  rw [this, hll]
+  rw [Nat.mod_eq_of_lt]
```
is therre a reason why these rewrites are not grouped into a single `rw`?

luisacicolini commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,642 @@
+  ext i hi
+  simp [BitVec.getElem_sshiftRight]
+  simp [show i ≤ w - 1 by omega]
+  simp [BitVec.getLsbD_signExtend]
```
```suggestion
  simp [BitVec.getElem_sshiftRight, show i ≤ w - 1 by omega, BitVec.getLsbD_signExtend]
```
if possible we should try to group things wihtin the same `simp` or `rw` (not always possible, so please do double check that this does not break the proof!) also, non-terminal simp

luisacicolini commented on SSA/Projects/RISCV64/Semantics2.lean:
```diff
@@ -0,0 +1,642 @@
+    rw [BitVec.toInt_signExtend_eq_toInt_bmod_of_le _ hw']
+  · simp only [not_le] at hw'
+    rw [BitVec.toInt_ofInt]
+    rw [BitVec.toInt_signExtend_of_le (by omega)]
```
can we group them?

luisacicolini reviewed (commented):
Thank you @salinhkuhn for the PR :)
I left a few style comments about the use of `simp only`, implicit arguments and grouping the application of rewrite rules/simp lemmas so that they're consistent with lean's style, could you please go over the bitvec proofs and make sure that they are consistent?

tobiasgrosser commented:
@salinhkuhn, as commented earlier please move everything about `BitVec` into the file `SSA/Projects/InstCombine/ForLean.lean`.

tobiasgrosser commented:
@salinhkuhn, I am also surprised you have redundant statements in `Semantics.lean` and `Semantics2.lean`. Is there a specific reason you decided to introduce redundancy here. I would have expected that this PR just adds additional `_bv` definitions plus proofs of equivalence into the file `Semantics.lean`. This would make this PR smaller and easier to review.

salinhkuhn committed:
Update SSA/Projects/RISCV64/Semantics2.lean

Co-authored-by: Luisa Cicolini <48860705+luisacicolini@users.noreply.github.com>

salinhkuhn commented:
> @salinhkuhn, I am also surprised you have redundant statements in `Semantics.lean` and `Semantics2.lean`. Is there a specific reason you decided to introduce redundancy here. I would have expected that this PR just adds additional `_bv` definitions plus proofs of equivalence into the file `Semantics.lean`. This would make this PR smaller and easier to review.

My intention was for this file to fully replace the current Semantics.lean file so that only one semantics file remains. However, I can adjust the PR if it's preferred to keep both files.

tobiasgrosser commented:
> > @salinhkuhn, I am also surprised you have redundant statements in `Semantics.lean` and `Semantics2.lean`. Is there a specific reason you decided to introduce redundancy here. I would have expected that this PR just adds additional `_bv` definitions plus proofs of equivalence into the file `Semantics.lean`. This would make this PR smaller and easier to review.
> 
> My intention was for this file to fully replace the current Semantics.lean file so that only one semantics file remains. However, I can adjust the PR if it's preferred to keep both files.

If this file should replace the current file, would it then make sense to do this in-place (by updating `Semantics.lean` or by just copying this file over `Semantics.lean`). Adding a separate file and then deleting the old one prevents git's history tracking support from working well. Also, the diff between the current definitions and the new ones is lost. So yes, please just keep updating `Semantics.lean`.

Commit:
started corrections but server keeps crashing

Commit:
fixed brackets and identation

Commit:
double check forLean file

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -89,171 +207,415 @@ def RTYPE_pure64_RISCV_XOR (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec
+
 -- wip: need to remove toInts and toNats
 def DIVW_pure64_signed (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec 64 :=
-  BitVec.signExtend 64
```
Anything from here and below seems to be an unintentional whitespace change that should be removed to reduce the diff.

tobiasgrosser commented on SSA/Projects/InstCombine/ForLean.lean:
```diff
@@ -454,15 +454,13 @@ theorem ofInt_neg_one : BitVec.ofInt w (-1) = -1#w := by
     simp
-  ·
-    simp only [Int.reduceNeg, ne_eq, h, not_false_eq_true, Nat.one_mod_two_pow_eq,
+  · simp only [Int.reduceNeg, ne_eq, h, not_false_eq_true, Nat.one_mod_two_pow_eq,
```
This change seems to be unrelated to the topic of this specific PR. I would remove it and submit as a separate cleanup PR.

salinhkuhn replied:
I fixed a syntax indentation mistake, so you want me to make a separate pr for this ?  I'm a bit confused

tobiasgrosser replied:
Right, you got two conflicting messages there.

1) this is a valuable cleanup

2) it is generally preferable to keep PRs focused on a single objective

The smaller a PR and the clearer it is focused on the objective in the commit message, the easier it is to review. In particular, when I review a PR, I ask at each line the question "why is this needed to achieve the goal in the commit message". In this particular instance it is relatively clear that this is an unrelated cleanup, so it is not too hindersome. However, it is not always obvious that something is unrelated and in some cases a bug may be introduced by something unrelated. You saw earlier how `git bisect` help to find bugs. It works a lot better if patches are focused and can in certain cases then help to identify if a bug was introduced in the main change of. a patch or in an unrelated refatoring. However, this only works if the unrelated refactoring is submitted in an independent PR.

It is a good habit to check if changes are unrelated and pull such changes out into their own PR. The change above is obviously beneficial. If you pull it out into a new PR, you get a thumbs-up almost immediately, and this change can be applied immediately. When I prepare a PR, I would typically try to commit such obvious unrelated changes independently (often committing before a review has happened and relying on people reviewing the already committed changes). The PRs that I then submit for review are then as small as possible to focus on the attention of the reviewers, saving everyone time.

Here a recent example of such a change: https://github.com/opencompl/lean-mlir/pull/1221. I also pulled this change out for you https://github.com/opencompl/lean-mlir/pull/1222, to show how I would do it and acknowledging that these early PR reviews are already quite draining: https://github.com/opencompl/lean-mlir/pull/1222.

salinhkuhn replied:
Okay thx for explaining, I will do that from now on.

tobiasgrosser commented on SSA/Projects/InstCombine/ForLean.lean:
```diff
@@ -454,15 +454,13 @@ theorem ofInt_neg_one : BitVec.ofInt w (-1) = -1#w := by
     norm_cast
-    simp only [Nat.cast_pow, Nat.cast_ofNat, Int.reduceNeg]
-    simp only [Int.reduceNegSucc, Int.reduceNeg]
+    simp only [Nat.cast_pow, Nat.cast_ofNat, Int.reduceNeg, Int.reduceNegSucc, Int.reduceNeg]
```
This change seems to be unrelated to the topic of this specific PR. I would remove it and together with the previous change as a separate cleanup PR.

salinhkuhn replied:
okay, yes I think I'm just confused about what I should have done here and what not . I did this change thanks to Luisa feedback, she advised me to merge to simps here if possible. I thought it's a good input and did it. I didn't change the logic of the proof or anything.

tobiasgrosser commented on SSA/Projects/InstCombine/ForLean.lean:
```diff
@@ -484,6 +482,83 @@ theorem sdiv_allOnes {w : ℕ} {x : BitVec w} :
+theorem setWidth_signExtend_eq_self {w w' : Nat} {x : BitVec w} (h : w ≤ w') : (x.signExtend w').setWidth w = x := by
+  ext i hi
+  simp  [hi, BitVec.getLsbD_signExtend]
+  omega
```
Thank you for these bitvec lemmas.

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -25,22 +32,86 @@ def UTYPE_pure64_AUIPC (imm : BitVec 20) (pc : BitVec 64) : BitVec 64 :=
+    BitVec.signExtend 64 (BitVec.shiftLeft (BitVec.extractLsb' 0 32 rs1_val) (shamt).toNat)
+
+def SHIFTIWOP_pure64_RISCV_SLLIW_bv (shamt : BitVec 5) (rs1_val : BitVec 64) : BitVec 64 :=
+    BitVec.signExtend 64 ((BitVec.extractLsb' 0 32 rs1_val)  <<< shamt)
```
Your new definitions use an indentation of four whitespace vs. two in the original code. Lean uses two whitespace, hence I would suggest to stick to two whitespace as well (throughout the file).

salinhkuhn replied:
done

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -25,22 +32,86 @@ def UTYPE_pure64_AUIPC (imm : BitVec 20) (pc : BitVec 64) : BitVec 64 :=
+    BitVec.signExtend 64 (BitVec.ushiftRight (BitVec.extractLsb' 0 32 rs1_val) (shamt).toNat)
+
+def SHIFTIWOP_pure64_RISCV_SRLIW_bv (shamt : BitVec 5) (rs1_val : BitVec 64) : BitVec 64 :=
+    BitVec.signExtend 64 ((BitVec.extractLsb' 0 32 rs1_val) >>> shamt)
```
These definitions are interesting. It seems you are **introducing** the `toNat` version, as you seemingly already had the pure version. Can you update the PR message to explain what you are doing here, e.g., "For some theorems, we previously already committed their pure variant but left out the `toNat` version. In this case, we now add the `toNat` version under the original theorem name. move the pure version to a `_bv` statement, and prove equivalence of both.

salinhkuhn replied:
thank you, yes that makes it clear.

tobiasgrosser reviewed (commented):
@salinhkuhn, this already looks amazing. There are a couple of comments that apply throughout the file (check if they apply anywhere else). After addressing those, there are likely only minor changes remaining (or none at all). Thank you for the fast turnaround.

Commit:
fix: whitespaces& identation

tobiasgrosser committed:
Merge remote-tracking branch 'origin/main' into toInt_toNat_removal_semantics

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -25,23 +32,87 @@ def UTYPE_pure64_AUIPC (imm : BitVec 20) (pc : BitVec 64) : BitVec 64 :=
+  SHIFTIWOP_pure64_RISCV_SLLIW shamt rs1_val = SHIFTIWOP_pure64_RISCV_SLLIW_bv shamt rs1_val :=
+  by
+  unfold SHIFTIWOP_pure64_RISCV_SLLIW SHIFTIWOP_pure64_RISCV_SLLIW_bv
+  rfl
```
```suggestion
theorem SHIFTIWOP_pure64_RISCV_SLLIW_eq_SHIFTIWOP_pure64_RISCV_SLLIW_bv (shamt : BitVec 5) (rs1_val : BitVec 64) :
    SHIFTIWOP_pure64_RISCV_SLLIW shamt rs1_val = SHIFTIWOP_pure64_RISCV_SLLIW_bv shamt rs1_val := by
  unfold SHIFTIWOP_pure64_RISCV_SLLIW SHIFTIWOP_pure64_RISCV_SLLIW_bv
  rfl
```

- theorem statement lines are typically indented by four spaces
- `by` typically follows right after the `:=`
Similarly, below.

tobiasgrosser commented on SSA/Projects/InstCombine/ForLean.lean:
```diff
@@ -477,6 +477,83 @@ theorem sdiv_allOnes {w : ℕ} {x : BitVec w} :
+    x.sshiftRight n =
+    ((x.signExtend (w + n)).extractLsb (w - 1 + n) n).setWidth w := by
+  ext i hi
+  simp [BitVec.getElem_sshiftRight, show i ≤ w - 1 by omega, BitVec.getLsbD_signExtend]
```
Non-terminal simp.

tobiasgrosser commented on SSA/Projects/InstCombine/ForLean.lean:
```diff
@@ -477,6 +477,83 @@ theorem sdiv_allOnes {w : ℕ} {x : BitVec w} :
+theorem toInt_toInt_ofInt_eq_toNat_toNat_ofNa {w w' : Nat } {x y : BitVec w} (h : w' ≤ w):
+    BitVec.ofNat w' (x.toNat * y.toNat) = BitVec.ofInt w' (x.toInt * y.toInt) := by
+  rw [BitVec.ofNat_mul]
+  simp
```
Non-terminal simp.

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -25,23 +32,87 @@ def UTYPE_pure64_AUIPC (imm : BitVec 20) (pc : BitVec 64) : BitVec 64 :=
+  unfold SHIFTIOP_pure64_RISCV_SLLI SHIFTIOP_pure64_RISCV_SLLI_bv
+  simp only [BitVec.shiftLeft_eq, BitVec.shiftLeft_eq']
+
+
```
```suggestion

```

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -53,21 +124,68 @@ def RTYPEW_pure64_RISCV_SUBW (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVe
+  = RTYPEW_pure64_RISCV_SRLW_bv rs2_val rs1_val := by
+  unfold RTYPEW_pure64_RISCV_SRLW RTYPEW_pure64_RISCV_SRLW_bv
+  simp only [extractLsb'_eq_setWidth, Nat.reduceLeDiff, BitVec.setWidth_setWidth_of_le,
+    BitVec.toNat_setWidth, Nat.reducePow, BitVec.ushiftRight_eq, BitVec.ushiftRight_eq']
```
A terminal simp does not need to be `simp only`. Drop the `only` and see if you can drop some of the arguments to this `simp` call.

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -89,83 +207,328 @@ def RTYPE_pure64_RISCV_XOR (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec
+  congr
+  apply BitVec.eq_of_toInt_eq
+  simp only [BitVec.toInt_extractLsb, Nat.shiftRight_zero, Nat.sub_zero, Nat.reduceAdd,
+    Nat.reducePow, BitVec.toInt_ofInt, Int.mul_bmod_bmod, Int.bmod_mul_bmod, BitVec.toInt_mul]
```
Drop the `only` at the end of the proof.

tobiasgrosser commented on SSA/Projects/RISCV64/Semantics.lean:
```diff
@@ -89,83 +207,328 @@ def RTYPE_pure64_RISCV_XOR (rs2_val : BitVec 64) (rs1_val : BitVec 64) : BitVec
+  have aa := BitVec.isLt rs1_val
+  have bb := BitVec.isLt rs2_val
+  have := @Nat.mul_lt_mul'' _ _ _ _ aa bb
+  simp only [Nat.reducePow, Nat.reduceMul] at this
```
Right before omega as the second-last tactic in a proof, you can (and should) also drop the `only`.

tobiasgrosser reviewed (approved):
Excellent. I left a couple of comments that are essential for code health. After these are addressed, I feel we can already merge this.

The proofs can likely be polished further, which would be useful but maybe is not on the critical path. If you feel like it, ask @luisacicolini for another proof-quality review, though.

Please merge yourself after you addressed the comment remaining comments and are happy with the overall patch.

Commit:
fix:non-terminal& terminal simp, whitespace

salinhkuhn commented:
> Excellent. I left a couple of comments that are essential for code health. After these are addressed, I feel we can already merge this.
> 
> The proofs can likely be polished further, which would be useful but maybe is not on the critical path. If you feel like it, ask @luisacicolini for another proof-quality review, though.
> 
> Please merge yourself after you addressed the comment remaining comments and are happy with the overall patch.

Yes, I will keep proof polishing on my to do list.
